CUDADataFormats/HGCal/interface/HGCConditions.h

```diff
@@ -0,0 +1,197 @@
+#ifndef CUDADataFormats_HGCal_HGCConditions_h
+#define CUDADataFormats_HGCal_HGCConditions_h
+
+class HeterogeneousHGCSiliconDetId {
+public:
+  constexpr HeterogeneousHGCSiliconDetId(uint32_t id) : id_(id) {}
+  constexpr int32_t type() { return (id_ >> kHGCalTypeOffset) & kHGCalTypeMask; }
+  constexpr int32_t zside() { return (((id_ >> kHGCalZsideOffset) & kHGCalZsideMask) ? -1 : 1); }
+  constexpr int32_t layer() { return (id_ >> kHGCalLayerOffset) & kHGCalLayerMask; }
+  constexpr int32_t waferUAbs() { return (id_ >> kHGCalWaferUOffset) & kHGCalWaferUMask; }
+  constexpr int32_t waferVAbs() { return (id_ >> kHGCalWaferVOffset) & kHGCalWaferVMask; }
+  constexpr int32_t waferU() {
+    return (((id_ >> kHGCalWaferUSignOffset) & kHGCalWaferUSignMask) ? -waferUAbs() : waferUAbs());
+  }
+  constexpr int32_t waferV() {
+    return (((id_ >> kHGCalWaferVSignOffset) & kHGCalWaferVSignMask) ? -waferVAbs() : waferVAbs());
+  }
+  constexpr int32_t waferX() { return (-2 * waferU() + waferV()); }
+  constexpr int32_t waferY() { return (2 * waferV()); }
+  constexpr int32_t cellU() { return (id_ >> kHGCalCellUOffset) & kHGCalCellUMask; }
+  constexpr int32_t cellV() { return (id_ >> kHGCalCellVOffset) & kHGCalCellVMask; }
+  constexpr int32_t nCellsSide() { return (type() == HGCalFine) ? HGCalFineN : HGCalCoarseN; }
+  constexpr int32_t cellX() {
+    const int32_t N = nCellsSide();
+    return (3 * (cellV() - N) + 2);
+  }
+  constexpr int32_t cellY() {
+    const int32_t N = nCellsSide();
+    return (2 * cellU() - (N + cellV()));
+  }
+
+private:
+  uint32_t id_;
+  enum waferType { HGCalFine = 0, HGCalCoarseThin = 1, HGCalCoarseThick = 2 };
+  static const int32_t HGCalFineN = 12;
+  static const int32_t HGCalCoarseN = 8;
+  static const int32_t kHGCalCellUOffset = 0;
+  static const int32_t kHGCalCellUMask = 0x1F;
+  static const int32_t kHGCalCellVOffset = 5;
+  static const int32_t kHGCalCellVMask = 0x1F;
+  static const int32_t kHGCalWaferUOffset = 10;
```

Comment:
this could be made a bit more robust with something like
```suggestion
  static const int32_t kHGCalCellUOffset = 0;
  static const int32_t kHGCalCellUWidth = 5;
  static const int32_t kHGCalCellUMask = (1 << kHGCalCellUWidth) - 1;
  static const int32_t kHGCalCellVOffset = kHGCalCellUOffset + kHGCalCellUWidth;
  static const int32_t kHGCalCellVWidth = 5;
  static const int32_t kHGCalCellVMask = (1 << kHGCalCellVWidth) - 1;
  static const int32_t kHGCalWaferUOffset = kHGCalCellVOffset + kHGCalCellVWidth;
```
*etc.*

that is, deriving both a field's mask and the next field's offset from a field's width.

This is just a suggestion: I think this is mostly up to the developer's taste, and whether the most robust approach makes sense depends if one foresees any updates to these structure in the future.

Reply:
using `constexpr` would be better.

Reply:
@fwyzard Here I merely copied the relevant data members from its [CPU counterpart](https://cmssdt.cern.ch/lxr/ident?_i=triggerCellV). Does it make sense to diverge? Besides, the overall structure of HGCal will potentially still change. Perhaps @rovere can comment.

Reply:
@Dr15Jones I get ```non-static data member '<var>' declared 'constexpr'``` if I replace ```static const``` by  ```constexpr```. This is actually expected, given that we are talking about members of a class. Please see [this](https://stackoverflow.com/questions/27408819/why-cant-non-static-data-members-be-constexpr). Are you instead proposing to get rid of the class?

Reply:
use `static constexpr`.

Reply:
I think the intention was `const` -> `constexpr` (keeping `static`), not `static const` -> `constexpr`.